From: Muqtadir <muqtadir5555@gmail.com>
Date: Sat, 02 Jul 2023 12:28:13 +0530
Subject: swap stylesheet: Swap GNOME-Shell's stylesheet to dark variant when dark style is toggled

Code taken from:
     ubuntu's patches: main-Support-loading-multiple-Yaru-theme-variants.patch
     upstream: bce07888099af5724b70be88094b48a490e322ba
---
--- a/js/ui/main.js
+++ b/js/ui/main.js
@@ -179,6 +179,7 @@
     sessionMode.connect('updated', _sessionUpdated);
 
     St.Settings.get().connect('notify::high-contrast', _loadDefaultStylesheet);
+    St.Settings.get().connect('notify::color-scheme', _loadDefaultStylesheet);
 
     // Initialize ParentalControlsManager before the UI
     ParentalControlsManager.getDefault();
@@ -427,6 +428,9 @@
     if (St.Settings.get().high_contrast)
         stylesheet = _getStylesheet(name.replace('.css', '-high-contrast.css'));
 
+    if (St.Settings.get().color_scheme === St.SystemColorScheme.PREFER_DARK)
+        stylesheet  = _getStylesheet(name.replace('.css', '-dark.css'));
+
     if (stylesheet == null)
         stylesheet = _getStylesheet(sessionMode.stylesheetName);
 
--- a/src/st/st-settings.h
+++ b/src/st/st-settings.h
@@ -25,9 +25,16 @@
 #define __ST_SETTINGS_H__
 
 #include <glib-object.h>
+#include <gdesktop-enums.h>
 
 G_BEGIN_DECLS
 
+typedef enum {
+  ST_SYSTEM_COLOR_SCHEME_DEFAULT = G_DESKTOP_COLOR_SCHEME_DEFAULT,
+  ST_SYSTEM_COLOR_SCHEME_PREFER_DARK = G_DESKTOP_COLOR_SCHEME_PREFER_DARK,
+  ST_SYSTEM_COLOR_SCHEME_PREFER_LIGHT = G_DESKTOP_COLOR_SCHEME_PREFER_LIGHT,
+} StSystemColorScheme;
+
 #define ST_TYPE_SETTINGS (st_settings_get_type ())
 G_DECLARE_FINAL_TYPE (StSettings, st_settings, ST, SETTINGS, GObject)
 
--- a/src/st/st-settings.c
+++ b/src/st/st-settings.c
@@ -26,6 +26,7 @@
 
 #include "st-private.h"
 #include "st-settings.h"
+#include "st-enum-types.h"
 
 #define KEY_ENABLE_ANIMATIONS     "enable-animations"
 #define KEY_PRIMARY_PASTE         "gtk-enable-primary-paste"
@@ -33,6 +34,7 @@
 #define KEY_FONT_NAME             "font-name"
 #define KEY_HIGH_CONTRAST         "high-contrast"
 #define KEY_GTK_ICON_THEME        "icon-theme"
+#define KEY_COLOR_SCHEME          "color-scheme"
 #define KEY_MAGNIFIER_ACTIVE      "screen-magnifier-enabled"
 #define KEY_DISABLE_SHOW_PASSWORD "disable-show-password"
 
@@ -44,6 +46,7 @@
   PROP_FONT_NAME,
   PROP_HIGH_CONTRAST,
   PROP_GTK_ICON_THEME,
+  PROP_COLOR_SCHEME,
   PROP_MAGNIFIER_ACTIVE,
   PROP_SLOW_DOWN_FACTOR,
   PROP_DISABLE_SHOW_PASSWORD,
@@ -71,6 +74,7 @@
   gboolean disable_show_password;
   gint drag_threshold;
   double slow_down_factor;
+  StSystemColorScheme color_scheme;
 };
 
 G_DEFINE_TYPE (StSettings, st_settings, G_TYPE_OBJECT)
@@ -185,6 +189,9 @@
     case PROP_GTK_ICON_THEME:
       g_value_set_string (value, settings->gtk_icon_theme);
       break;
+    case PROP_COLOR_SCHEME:
+      g_value_set_enum (value, settings->color_scheme);
+      break;
     case PROP_MAGNIFIER_ACTIVE:
       g_value_set_boolean (value, settings->magnifier_active);
       break;
@@ -276,6 +283,18 @@
                                                     ST_PARAM_READABLE);
 
   /**
+   * StSettings:color-scheme:
+   *
+   * The preferred color-scheme
+   */
+  props[PROP_COLOR_SCHEME] = g_param_spec_enum ("color-scheme",
+                                                "Color scheme",
+                                                "Color scheme",
+                                                ST_TYPE_SYSTEM_COLOR_SCHEME,
+                                                ST_SYSTEM_COLOR_SCHEME_DEFAULT,
+                                                ST_PARAM_READABLE);
+
+  /**
    * StSettings:magnifier-active:
    *
    * Whether the accessibility magnifier is active.
@@ -339,6 +358,12 @@
       g_object_notify_by_pspec (G_OBJECT (settings),
                                 props[PROP_GTK_ICON_THEME]);
     }
+  else if (g_str_equal (key, KEY_COLOR_SCHEME))
+    {
+      settings->color_scheme = g_settings_get_enum (g_settings, key);
+      g_object_notify_by_pspec (G_OBJECT (settings),
+                                props[PROP_COLOR_SCHEME]);
+    }
 }
 
 static void
@@ -422,6 +447,8 @@
                                                KEY_FONT_NAME);
   settings->gtk_icon_theme = g_settings_get_string (settings->interface_settings,
                                                     KEY_GTK_ICON_THEME);
+  settings->color_scheme = g_settings_get_enum (settings->interface_settings,
+                                                KEY_COLOR_SCHEME);
   settings->drag_threshold = g_settings_get_int (settings->mouse_settings,
                                                  KEY_DRAG_THRESHOLD);
   settings->magnifier_active = g_settings_get_boolean (settings->a11y_applications_settings,
